src/MLIAP/mliap_so3.cpp

```diff
@@ -0,0 +1,1422 @@
+    m_rootpq[i] = 0.0;
+
+  for (int p = 1; p < m_ldim; p++)
+    for (int q = 1; q < m_ldim; q++)
+      m_rootpq[p * m_ldim + q] = sqrt(static_cast<double>(p) / q);
+
+  memory->create(m_idxu_block, m_ldim, "MLIAP_SO3:idxu_bloc");
+
+  for (i = 0; i < m_ldim; i++)
+    m_idxu_block[i] = 0;
+
+  totali = pow(m_lmax + 2, 2);
+  memory->create(m_idxylm, totali, "MLIAP_SO3:idxylm");
+
+  for (i = 0; i < totali; i++)
+    m_idxylm[i] = 0;
+
+  m_idxu_count = 0, m_idxy_count = 0;
+
+  for (int l = 0; l < m_ldim; l++) {
+    m_idxu_block[l] = m_idxu_count;
+    for (int mb = 0; mb < l + 1; mb++)
+      for (int ma = 0; ma < l + 1; ma++) {
+        if (l % 2 == 0 && ma == l / 2) {
+          m_idxylm[m_idxy_count] = m_idxu_count;
+          m_idxy_count += 1;
+        }
+        m_idxu_count += 1;
+      }
+  }
+
+  m_numYlms = (m_lmax + 1) * (m_lmax + 1);
+}
+
+void MLIAP_SO3::init_arrays(int natoms, int *numneighs, int ncoefs)
+{
+  int totali = natoms * ncoefs;
+  memory->destroy(m_plist_r);
+  memory->create(m_plist_r, totali, "MLIAP_SO3:m_plist_r");
+  memory->destroy(m_plist_i);
+  memory->create(m_plist_i, totali, "MLIAP_SO3:m_plist_i");
+
+  totali = m_nmax * m_numYlms;
+  memory->destroy(m_clist_r);
+  memory->create(m_clist_r, totali, "MLIAP_SO3:m_clist_r");
+  memory->destroy(m_clist_i);
+  memory->create(m_clist_i, totali, "MLIAP_SO3:m_clist_i");
+
+  totali = m_idxu_count;
+  memory->destroy(m_ulist_r);
+  memory->create(m_ulist_r, totali, "MLIAP_SO3:m_ulist_r");
+  memory->destroy(m_ulist_i);
+  memory->create(m_ulist_i, totali, "MLIAP_SO3:m_ulist_i");
+
+  totali = (m_lmax + 2) * (m_lmax + 2);
+  memory->destroy(m_Ylms_r);
+  memory->create(m_Ylms_r, totali, "MLIAP_SO3:m_Ylms_r");
+  memory->destroy(m_Ylms_i);
+  memory->create(m_Ylms_i, totali, "MLIAP_SO3:m_Ylms_i");
+
+  totali = (m_lmax + 1) * (m_lmax + 1) * 3;
+  memory->destroy(m_dYlm_r);
+  memory->create(m_dYlm_r, totali, "MLIAP_SO3:m_dYlm_r");
+  memory->destroy(m_dYlm_i);
+  memory->create(m_dYlm_i, totali, "MLIAP_SO3:m_dYlm_i");
+
+  totali = m_nmax * m_numYlms * 3;
+  memory->destroy(m_dclist_r);
+  memory->create(m_dclist_r, totali, "MLIAP_SO3:m_dclist_r");
+  memory->destroy(m_dclist_i);
+  memory->create(m_dclist_i, totali, "MLIAP_SO3:m_dclist_i");
+
+  totali = 3 * m_nmax * (m_nmax + 1) * (m_lmax + 1) / 2;
+  memory->destroy(m_tempdp_r);
+  memory->create(m_tempdp_r, totali, "MLIAP_SO3:m_tempdp_r");
+
+  totali = m_nmax * m_numYlms;
+  memory->destroy(m_clisttot_r);
+  memory->create(m_clisttot_r, totali, "MLIAP_SO3:m_clisttot_r");
+  memory->destroy(m_clisttot_i);
+  memory->create(m_clisttot_i, totali, "MLIAP_SO3:m_clisttot_i");
+  m_init_arrays = 1;
+}
+
+double MLIAP_SO3::memory_usage()
+{
+  double bytes;
+
+  bytes = 0;
+  bytes += (double) m_nmax * m_nmax * sizeof(double);
+
+  return bytes;
+}
+
+void MLIAP_SO3::swap(double *a, double *b)
+{
+  double t = *a;
+  *a = *b;
+  *b = t;
+}
+
+int MLIAP_SO3::partition(double arr[], int low, int high,
```

Comment:
The Jacobi class included in LAMMPS can return sorted Eigenvalues directly, IIRC. So this would be superfluous.